crates/core/component/funding/src/component/state_key.rs

```diff
@@ -15,78 +15,247 @@ pub mod lqt {
-
-                bytes_prefix.copy_from_slice(&prefix(epoch_index));
-                bytes_asset.copy_from_slice(&asset.to_bytes());
-                bytes_power.copy_from_slice(&((!power).to_be_bytes()));
-                bytes_voter.copy_from_slice(&voter.to_vec());
-
-                bytes
+            pub mod by_asset {
+                use super::*;
+
+                const PART1: &'static str = "/by_asset/";
+
+                pub mod total {
+                    use super::*;
+
+                    const PART2: &'static str = "total/";
+                    const KEY_LEN: usize =
+                        PART0.len() + EPOCH_LEN + PART1.len() + PART2.len() + ASSET_LEN;
+
+                    pub(crate) fn key(epoch_index: u64, asset: asset::Id) -> [u8; KEY_LEN] {
+                        let mut bytes = [0u8; KEY_LEN];
+
+                        let rest = &mut bytes;
+                        let (bytes_part0, rest) = rest.split_at_mut(PART0.len());
+                        let (bytes_epoch, rest) = rest.split_at_mut(EPOCH_LEN);
+                        let (bytes_part1, rest) = rest.split_at_mut(PART1.len());
+                        let (bytes_part2, rest) = rest.split_at_mut(PART2.len());
+                        let (bytes_asset, _) = rest.split_at_mut(ASSET_LEN);
+
+                        bytes_part0.copy_from_slice(PART0.as_bytes());
+                        bytes_epoch.copy_from_slice(format_epoch(epoch_index).as_bytes());
+                        bytes_part1.copy_from_slice(PART1.as_bytes());
+                        bytes_part2.copy_from_slice(PART2.as_bytes());
+                        bytes_asset.copy_from_slice(asset.to_bytes().as_slice());
+
+                        bytes
+                    }
+                }
+
+                pub mod ranked {
+                    use super::*;
+
+                    const PART2: &'static str = "ranked/";
+                    const PREFIX_LEN: usize = PART0.len() + EPOCH_LEN + PART1.len() + PART2.len();
+
+                    pub(crate) fn prefix(epoch_index: u64) -> [u8; PREFIX_LEN] {
+                        let mut bytes = [0u8; PREFIX_LEN];
+
+                        let rest = &mut bytes;
+                        let (bytes_part0, rest) = rest.split_at_mut(PART0.len());
+                        let (bytes_epoch, rest) = rest.split_at_mut(EPOCH_LEN);
+                        let (bytes_part1, rest) = rest.split_at_mut(PART1.len());
+                        let (bytes_part2, _) = rest.split_at_mut(PART2.len());
+
+                        bytes_part0.copy_from_slice(PART0.as_bytes());
+                        bytes_epoch.copy_from_slice(format_epoch(epoch_index).as_bytes());
+                        bytes_part1.copy_from_slice(PART1.as_bytes());
+                        bytes_part2.copy_from_slice(PART2.as_bytes());
+
+                        bytes
+                    }
+
+                    const KEY_LEN: usize = PREFIX_LEN + POWER_LEN + ASSET_LEN;
```

Comment:
nit: `EPOCH_PREFIX_LEN` so we can read the schema in one-line. i think that's my main feedback. the code is more robust, if we have a single place to read the schema, it's perfect